src/librustc_mir/build/block.rs

```diff
@@ -19,11 +19,18 @@ impl<'a,'tcx> Builder<'a,'tcx> {
                      mut block: BasicBlock,
                      ast_block: &'tcx hir::Block)
                      -> BlockAnd<()> {
-        let this = self;
-        let Block { extent, span: _, stmts, expr } = this.hir.mirror(ast_block);
-        this.in_scope(extent, block, |this| {
+        let Block { extent, span, stmts, expr } = self.hir.mirror(ast_block);
+        self.in_scope(extent, block, move |this| {
             unpack!(block = this.stmts(block, stmts));
-            this.into(destination, block, expr)
+            match expr {
+                Some(expr) => this.into(destination, block, expr),
```

Comment:
Do we still need the impl of `EvalInto` for `Option<ExprRef<'tcx>>`? I suspect not. In any case I think perhaps we ought to remove it, since it's ambiguous what should happen on `None`.